File: src/test/java/cz/kb/oleg/strategy/events/EventTwoStrategy.java
package cz.kb.oleg.strategy.events;

import cz.kb.oleg.strategy.events.dto.EventDto;
import cz.kb.oleg.strategy.events.dto.EventTwoDto;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class EventTwoStrategy extends EventStrategy<EventDto> {

    @Override
    @NonNull
    public Class<EventTwoDto> getApplicableClass() {
        return EventTwoDto.class;
    }

    @Override
    public void apply(EventDto dto) {
        if (!(dto instanceof EventTwoDto eventTwoDto)) {
            throw new IllegalArgumentException("Invalid DTO type: " + dto.getClass().getName());
        }
        log.info("Handled by EventTwoStrategy: {}", eventTwoDto);
    }

}

File: src/test/kotlin/com/winterhate/strategy/events/EventTwoStrategy.kt
package com.winterhate.strategy.events

import com.winterhate.strategy.events.dto.EventDto
import com.winterhate.strategy.events.dto.EventTwoDto
import org.slf4j.Logger
import org.springframework.stereotype.Component

@Component
class EventTwoStrategy(
    private val log: Logger
) : EventStrategy<EventDto> {

    override fun getApplicableClass(): Class<out EventDto> = EventTwoDto::class.java

    override fun apply(data: EventDto) {
        val eventTwoDto = data as? EventTwoDto
            ?: throw IllegalArgumentException("Invalid DTO type: ${data::class.java.name}")
        log.info("Handled by EventTwoStrategy: {}", eventTwoDto)
    }

}

